File: CircularLinkedList.java
public class CircularLinkedList {
    private Node head;
    private Node tail;

    public CircularLinkedList() {
        this.head = null;
        this.tail = null;
    }

    public boolean isEmpty() {
        return head == null;
    }

    public void insertAtBeginning(int data) {
        Node newNode = new Node(data);

        if (isEmpty()) {
            head = newNode;
            tail = newNode;
            newNode.next = newNode;
        } else {
            newNode.next = head;
            head = newNode;
            tail.next = newNode;
        }
    }

    public void insertAtEnd(int data) {
        Node newNode = new Node(data);

        if (isEmpty()) {
            head = newNode;
            tail = newNode;
            newNode.next = newNode;
        } else {
            newNode.next = tail.next;
            tail.next = newNode;
            tail = newNode;
        }
    }

    public void deleteAtBeginning() {
        if (isEmpty()) {
            return;
        }

        if (head == tail) {
            head = null;
            tail = null;
        } else {
            head = head.next;
            tail.next = head;
        }
    }

    public void deleteAtEnd() {
        if (isEmpty()) {
            return;
        }

        if (head == tail) {
            head = null;
            tail = null;
        } else {
            Node currentNode = head;
            while (currentNode.next != tail) {
                currentNode = currentNode.next;
            }

            tail = currentNode;
            tail.next = head;
        }
    }

    public void printList() {
        Node currentNode = head;

        while (currentNode != null) {
            System.out.print(currentNode.data + " ");
            currentNode = currentNode.next;

            if (currentNode == head) {
                break;
            }
        }

        System.out.println();
    }

    private class Node {
        private int data;
        private Node next;

        public Node(int data) {
            this.data = data;
            this.next = null;
        }
    }
}

File: DSA in Python/circular_linkedlist.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class CircularLinkedList:
    def __init__(self):
        self.head = None

    def insert(self, val, k):
        new_node = Node(val)
        if self.head is None:
            # List is empty, initialize with the new node
            self.head = new_node
            new_node.next = new_node
        elif k == 1:
            # Insert at the beginning
            temp = self.head
            while temp.next != self.head:
                temp = temp.next
            new_node.next = self.head
            self.head = new_node
            temp.next = self.head
        else:
            # Insert at position k
            temp = self.head
            while (k - 2) > 0 and temp.next != self.head:
                temp = temp.next
                k -= 1
            new_node.next = temp.next
            temp.next = new_node

    def delete(self, k):
        if self.head is None:
            print("List is empty")
            return
        if k == 1:
            # Delete the first node
            temp = self.head
            tail = self.head
            while tail.next != self.head:
                tail = tail.next
            if self.head.next == self.head:
                # Only one node in the list
                self.head = None
            else:
                self.head = self.head.next
                tail.next = self.head
            print(f"Deleted: {temp.data}")
            del temp
        else:
            # Delete node at position k
            temp = self.head
            prev = None
            while (k - 1) > 0 and temp.next != self.head:
                prev = temp
                temp = temp.next
                k -= 1
            prev.next = temp.next
            print(f"Deleted: {temp.data}")
            del temp

    def print_list(self):
        if self.head is None:
            print("List is empty")
            return
        temp = self.head
        while True:
            print(temp.data, end=" ")
            temp = temp.next
            if temp == self.head:
                break
        print()

# Example usage
cll = CircularLinkedList()
print("hello\n")
cll.insert(111, 1)
cll.insert(2322, 2)
cll.insert(3333, 3)
cll.print_list()
cll.insert(4444, 2)
cll.print_list()
cll.delete(1)
print("Final list: ")
cll.print_list()
